File: client/jade/jade_object.rb
# frozen_string_literal: true

require './client/packet'
require './jade/jade_pair'
require './jade/jade_exceptions'

# The JadeObject is a an object that stores an array of JadePairs.
# JadeObjects can be serialized to strings and files making.
# They can also be deserialized from strings and files.
# Additionally if the JadeObject contains the correct parameters it can be converted to another object.
# In this case JadeObject can be converted to object type Packet.
#
# @author Joseph Beck
# @!visibility public
class JadeObject

  public

  # Args are used outside of the object itself require an attr accessor.
  #
  # @!attribute :args
  # @!visibility public
  attr_accessor :args

  # The initializer of the JadeObject just takes in an array of JadePairs.
  # JadePairs are essential to the functionality of the JadeObject.
  #
  # @param [JadePair[]]
  # @return [NilClass]\
  # @!visibility public
  def initialize(args)
    @args = args
  end

  # to_packet constructs a packet with args as long as args are valid for a packet conversion.
  # Being able to convert a JadeObject to an object of type Packet allows for great flexibility within the program.
  #
  # @param [NilClass]
  # @return [Packet or NilClass]
  # @!visibility public
  def to_packet
    make_packet(@args) if valid_packet_conversion?
  end

  # to_str converts the current JadeObject to a string.
  # This provides an easy way to view the current state of a JadeObject.
  #
  # @param [String]
  # @return [NilClass]
  # @!visibility public
  def to_str
    output = ''

    @args.each_with_index do |arg, index|
      output += arg.to_str
      output += "\n" if index != @args.length - 1
    end
    output
  end

  # to_array converts the current instance of the JadeObject to an array.
  # The array is built of each arg of the JadeObject converted to a string.
  #
  # @param [NilClass]
  # @return [String[]]
  # @!visibility public
  def to_array
    args_values = []
    @args.each do |arg|
      args_values << arg.to_str
    end
  end

  # Checks whether the current args of this instance of the JadeObject is nil?.
  #
  # @param [NilClass]
  # @return [Boolean]
  # @!visibility public
  def empty?
    @args.nil?
  end

  private

  # Make packet constructs an object of type Packet from supplied args.
  # This method is called from the to_packet method.
  #
  # @param [JadePair[]]
  # @return [Packet]
  # @!visibility public
  def make_packet(args)
    sender = args[0].second
    reciever = args[1].second
    type = args[2].second
    data = args[3].second
    check_sum = args[4].second
    Packet.new(sender, reciever, type, data, check_sum)
  end

  # Checks whether the current instance of the JadeObject.
  # It does this by checking the equality of the first of the item of args.
  #
  # @param [NilClass]
  # @return [Boolean]
  # @!visibility public
  def valid_packet_conversion?
    return false unless @args.length.eql? 5

    ((@args[0].first.eql? 'sender') &&
     (@args[1].first.eql? 'reciever') &&
     (@args[2].first.eql? 'type') &&
     (@args[3].first.eql? 'data') &&
     (@args[4].first.eql? 'checkSum'))
  end
end

File: robot/src/jade/JadeObject.java
package jade;

import client.Packet;

/**
 * The JadeObject class implements the object JadeObject which stores an array of JadePairs.
 * JadeObjects are able to be serialized and deserialized to strings by the respective tools.
 * JadeObjects can also be converted to Packets in order to send data in a formatted way over a TCP Server.
 * 
 * @author  Joseph Beck
 * @version 0.1
 * @since   2023
 */
public class JadeObject {

    /**
     * Stores an array of JadePairs as args.
     */
    private JadePair[] args;

    /**
     * Constructor of the Jade Object, initialises args.
     * 
     * @param JadePair[] arguments
     * @return none
     */
    public JadeObject(JadePair[] args) {
        this.args = args;
    }

    /**
     * This method converts the current jade object to a new object of type Packet.
     * 
     * @return Packet
     * @exception JadeConversionException
     */
    public Packet toPacket() {
        try {
            if (validateArgs()) {
                return makePacket();
            } else {
                throw new JadeConversionException("args length = " + args.length);
            }
        } catch (JadeConversionException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * This method, makePacket, makes a new object of type Packet from the args (JadePairs) of this object.
     * 
     * @return Packet
     */
    private Packet makePacket() {
        return new Packet(  args[0].getSecond(), 
                            args[1].getSecond(),
                            args[2].getSecond(),
                            args[3].getSecond(),
                            args[4].getSecond());
    }

    /**
     * This method, validateArgs, is used to check whether the current JadeObject is able to be converted to type Packet.
     * 
     * @return boolean
     */
    private boolean validateArgs() {
        try {
            if (args.length == 5) {
                return (
                    args[0].getFirst().equals("sender") && 
                    args[1].getFirst().equals("reciever") &&
                    args[2].getFirst().equals("type") &&
                    args[3].getFirst().equals("data") &&
                    args[4].getFirst().equals("checkSum"));
            } 
            throw new JadeValidationException("args length of " + args.length + " does not satisfy the requirements");
        } catch (JadeValidationException | IndexOutOfBoundsException e) {
            e.printStackTrace();
            return false;
        } 
    }   

    /**
     * This method check whether the args of the Jade Object is empty, true for empty false for not.
     * 
     * @return boolean
     */
    public boolean isEmpty() { 
        return args.length == 0; 
    }

    /**
     * getArgs returns the JadePair array that stores the arguments of the Jade object.
     * 
     * @return JadePair[] args
     */
    public JadePair[] getArgs() { 
        return this.args; 
    }

    /**
     * This method returns a string representation of the Jade Object.
     * 
     * @return String
     */
    @Override
    public String toString() {
        String output = "";

        for (int i = 0; i < args.length; i++) {
            output += (i < args.length-1) ? args[i] + "\n" : args[i]; 
        }

        return output;
    }
}